KEYCLOAK-19028 KEYCLOAK-19029 Add HotRod Map storage implementation

mhajas opened the pull request:
<!---
Please read https://github.com/keycloak/keycloak/blob/master/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

mhajas committed:
KEYCLOAK-19028 Add HotRod Map storage implementation

hmlnarik commented on model/map/src/main/java/org/keycloak/models/map/storage/chm/ConcurrentHashMapCRUDOperations.java:
```diff
@@ -0,0 +1,85 @@
+
+import static org.keycloak.utils.StreamsUtil.paginatedStream;
+
+public interface ConcurrentHashMapCRUDOperations<V extends AbstractEntity & UpdatableEntity, M> {
```
```suggestion
public interface ConcurrentHashMapCrudOperations<V extends AbstractEntity & UpdatableEntity, M> {
```

hmlnarik commented on testsuite/integration-arquillian/tests/pom.xml:
```diff
@@ -1414,6 +1423,80 @@
         </profile>
 
+        <profile>
+            <id>hot-rod-store</id>
```
```suggestion
            <id>map-storage-hot-rod</id>
```

Usage of this profile should be documented in HOW-TO-RUN.md

hmlnarik reviewed (commented):
Very nice, @mhajas! Just a couple of minor comments, please see inline.
Also note that this may clash with #8678.

martin-kanis commented on model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/HotRodMapStorage.java:
```diff
@@ -0,0 +1,178 @@
+        String queryString = "SELECT id " + iqmcb.getIckleQuery();
+
+        if (!queryParameters.getOrderBy().isEmpty()) {
+            queryString += " ORDER BY " + queryParameters.getOrderBy().stream().map(HotRodMapStorage::toOrderString)
```
Is ordering required with delete operation?

mhajas replied:
I think should do it. In the future, we may want to have some default value for `maxResult` to prevent the removal of too many items and we need to remove always the same values. We are sorting also in [CHM](https://github.com/keycloak/keycloak/blob/main/model/map/src/main/java/org/keycloak/models/map/storage/chm/ConcurrentHashMapStorage.java#L145).

martin-kanis commented on model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/HotRodMapStorage.java:
```diff
@@ -0,0 +1,178 @@
+        String queryString = iqmcb.getIckleQuery();
+
+        if (!queryParameters.getOrderBy().isEmpty()) {
+            queryString += " ORDER BY " + queryParameters.getOrderBy().stream().map(HotRodMapStorage::toOrderString)
```
Is ordering required with count operation?

mhajas replied:
This is a good question. We probably don't need it. 

However, there is another thing similar to this for this method. Currently, we are ignoring pagination for `getCount` as [`hitCount`](https://docs.jboss.org/infinispan/11.0/apidocs/org/infinispan/query/dsl/QueryResult.html#hitCount()) method ignores it. We use it for CHM though.

I am not sure whether we need to have pagination for `getCount`, @martin-kanis  @hmlnarik WDYT?

hmlnarik replied:
For `getCount`, the pagination should be ignored, as the number of the pages is derived from the count

mhajas replied:
Thanks @hmlnarik, in that case, the version of `getCount` that is in this PR should be correct.

martin-kanis reviewed (changes requested):
Great work @mhajas! I added some questions/comments. Could you also add copyrights to every new file please?

mhajas commented:
Thank you for reviews @hmlnarik and @martin-kanis. I will have a look soon.

Pipeline runs seem to be ok (I cannot see any additional failures caused by these changes):
https://master-jenkins.redhat.com/job/universal-test-pipeline-server/900/
https://master-jenkins.redhat.com/job/universal-test-pipeline-adapters/714/

mhajas commented on model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/IckleQueryMapModelCriteriaBuilder.java:
```diff
@@ -0,0 +1,133 @@
+    }
+
+    @Override
+    public ModelCriteriaBuilder<M> and(ModelCriteriaBuilder<M>... builders) {
```
`or` and `and` methods are probably wrong, I need to take another look and fix them according to the Javadoc

mhajas commented:
@hmlnarik @martin-kanis I updated PR based on your comments and I also rebased it to the latest main.

Please notice the last change: https://github.com/keycloak/keycloak/compare/beacc12360865e8d30bed0be6d1343f4aa000e61..c2eacd9ed880ee611cbe8e41bd2347d2e131df3d

We had a bug in the `IckleCriteriaBuilder` when using `and` method and more `builders` that contains the same named-attribute where we need to resolve conflicts. This happens for example here: https://github.com/keycloak/keycloak/blob/main/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/client/ClientSearchTest.java#L112

martin-kanis reviewed (dismissed):
Looks good! Thank you Michal.

hmlnarik commented on model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/HotRodMapStorageProviderFactory.java:
```diff
@@ -0,0 +1,102 @@
+
+    @Override
+    public boolean isSupported() {
+        return true;
```
Apologies for late spot. This should be supported if and only map-storage feature is enabled. See e.g. https://github.com/keycloak/keycloak/blob/36da2d20e97a10821c7305881d3ffa343b5ad001/model/map/src/main/java/org/keycloak/models/map/storage/chm/ConcurrentHashMapStorageProviderFactory.java#L341

hmlnarik reviewed (approved):
Thanks @martin-kanis and @mhajas!
